tests/src/whisk/core/database/test/MultipleReadersSingleWriterCacheTests.scala

```diff
@@ -0,0 +1,215 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.database.test
+
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+import scala.util.Failure
+import scala.util.Success
+
+import org.scalatest.FlatSpec
+import org.scalatest.Matchers
+
+import akka.actor.ActorSystem
+import whisk.common.Logging
+import whisk.common.TransactionId
+import whisk.core.database.MultipleReadersSingleWriterCache
+
+class MultipleReadersSingleWriterCacheTests(nIters: Int = 3) extends FlatSpec
+    with Matchers
+    with MultipleReadersSingleWriterCache[String, String]
+    with Logging {
+
+    "the cache" should "support simple CRUD" in {
+        System.out.println();
+        System.out.println("simple CRUD");
+
+        val inhibits = doReadWriteRead("foo").go(0)
+        inhibits.debug(this)
+
+        inhibits.nReadInhibits.get should be(0)
+        cacheSize should be(1)
+    }
+
+    "the cache" should "support concurrent CRUD to different keys" in {
+        //
+        // for the first iter, all reads are not-cached and each thread
+        // requests a different key, so we expect no read inhibits, and a
+        // bunch of write inhibits
+        //
+        val inhibits = doCRUD("CONCURRENT CRUD to different keys", { i => "foop_" + i })
+        inhibits.nReadInhibits.get should be(0)
+        inhibits.nWriteInhibits.get should not be (0)
+
+        //
+        // after the first iter, the keys already exist, so the first read
+        // should be cached, resulting in the writes proceeding more
+        // smoothly this time, thus inhibiting some of the second reads
+        //
+        for (i <- 1 to nIters - 1) {
+            doCRUD("CONCURRENT CRUD to different keys", { i => "foop_" + i })
+                .nReadInhibits.get should not be (0)
+        }
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys" in {
+        for (i <- 1 to nIters) {
+            doCRUD("CONCURRENT CRUD to shared keys", sharedKeys)
+                .nWriteInhibits.get should not be (0)
+        }
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys (zero latency)" in {
+        var someInhibits = false
+        for (i <- 1 to nIters) {
+            someInhibits = doCRUD("concurrent CRUD to shared keys (zero latency)", sharedKeys, 0)
+                .someInhibits
+        }
+        someInhibits should not be (false)
```

Comment:
overall, i think this test suite is of low value. i implemented it initially just to help me catch glaring bugs. the reality is that, the entire point of this cache is to catch race conditions. it is impossible to generate an interleaving that, in a fixed amount of time, is guaranteed to trigger (or not trigger) a race. 

so we need to decide what to do with this test. i would suggest that we leave it disabled for normal runs, and only use it when further developing (if we ever do) the cache.